scripts/linters/pylint_extensions.py

```diff
@@ -1960,6 +2059,427 @@ def process_tokens(self, tokens):
                         line=line_num)
 
 
+class ExceptionalTypesCommentChecker(checkers.BaseChecker):
+    """Custom pylint checker which checks that there is always a comment
+    for exceptional types in the backend type annotations.
+    """
+
+    EXCEPTIONAL_TYPE_STATUS_DICT = {
+        'type_comment_present': False,
+        'type_comment_line_num': 0,
+        'is_function_signature_block': True,
+        'outside_args_section': True,
+        'type_present_inside_arg_section': False,
+        'type_present_inside_return_section': False,
+        'type_present_in_function_signature': False,
+        'args_section_end_line_num': 0,
+        'func_def_start_line': 0,
+    }
+
+    EXCLUDED_DIRS_HAVING_EXCEPTIONAL_TYPE_COMMENTS = copy.deepcopy(
+        EXCLUDED_TYPE_COMMENT_DIRECTORIES
+    )
+
+    __implements__ = interfaces.IAstroidChecker
+
+    name = 'comment-for-exceptional-types'
+    priority = -1
+    msgs = {
+        'C0047': (
+            'Any type is used. Please try to use a more specific type instead.',
+            'any-type-used',
+            'Annotations with Any type should only be done for exceptional'
+            ' cases with proper comments.'
+        ),
+        'C0048': (
+            'cast function is used. Please add a proper comment if'
+            ' cast is needed.',
+            'cast-func-used',
+            'Casting of any value should be done with a proper comment.'
+        ),
+        'C0049': (
+            'object class is used. Please try to use a more specific'
+            ' type instead.',
+            'object-class-used',
+            'Annotations with object should only be done for exceptional'
+            ' cases with proper comments.'
+        )
+    }
+
+    def visit_module(self, node):
+        """Visit a module to ensure that there is a comment for each exceptional
+        type (cast, Any and object).
+
+        Args:
+            node: astroid.scoped_nodes.Module. Node to access module content.
+        """
+        module_excluded_from_comment_checks = False
+        for directory in self.EXCLUDED_DIRS_HAVING_EXCEPTIONAL_TYPE_COMMENTS:
+            if re.search(directory, node.root().file):
```

Comment:
ditto: `re.search` doesn't seem correct here since `directory` is really just a path, not a regex

Reply:
Substring search is used